security/pkg/nodeagent/secretfetcher/secretfetcher.go

```diff
@@ -195,90 +195,161 @@ func isIngressGatewaySecret(scrt *v1.Secret) bool {
 	if len(cert) > 0 && len(key) > 0 {
 		certAndKeyExist = true
 	}
-	return cert, key, root, certAndKeyExist
+
+	return cert, key, certAndKeyExist
 }
 
-func (sf *SecretFetcher) scrtAdded(obj interface{}) {
-	scrt, ok := obj.(*v1.Secret)
-	if !ok {
-		secretFetcherLog.Warnf("Failed to convert to secret object: %v", obj)
-		return
+// extractCACert extracts the client CA certificate from either the Compound
+// Secret, or from a separate Kubernetes TLS secret that has CA cert in `tls.crt` field.
+func extractCACert(scrt *v1.Secret, fromCompoundSecret bool) (caCert []byte, exist bool) {
+	if len(scrt.Data[genericScrtCaCert]) > 0 {
+		caCert = scrt.Data[genericScrtCaCert]
+	} else if !fromCompoundSecret {
+		caCert = scrt.Data[tlsScrtCert]
 	}
 
+	return caCert, len(caCert) > 0
+}
+
+// extractK8sSecretIntoSecretItem extracts a server cert/key pair and a client CA
+// certificate from the k8s Secret into a pair of SecretItems. Returns SecretItems and a boolean
+// indicating whether this is a CA only k8s Secret.
+// A CA only k8s secret has name suffix `-cacert`, and is ONLY considered for a client CA;
+// either a `cacert` or `tls.crt` must be provided.
+// Otherwise the Secret can hold a server cert/key pair in `tls.crt`/`tls.key`,
+// or a server cert/key pair in `cert`/`key` and an optional client CA cert in
+// `-cacert`. A Secret with server cert/key and client CA cert is considered as a compound secret.
+func extractK8sSecretIntoSecretItem(scrt *v1.Secret, t time.Time) (serverItem, clientCAItem *model.SecretItem, isCAOnlySecret bool) {
 	resourceName := scrt.GetName()
-	if !isIngressGatewaySecret(scrt) {
-		secretFetcherLog.Debugf("secret %s is not an ingress gateway secret, skip adding secret", resourceName)
-		return
-	}
+	isCAOnlySecret = strings.HasSuffix(resourceName, IngressGatewaySdsCaSuffix)
+
+	// Extract CA cert from CA only k8s secret.
+	if isCAOnlySecret {
+		caCert, exist := extractCACert(scrt, false /* fromCompoundSecret */)
+		if !exist {
+			secretFetcherLog.Warnf("failed load CA only secret from %s: no 'cacert' or 'tls.crt' key in the secret", resourceName)
+			return nil, nil, isCAOnlySecret
+		}
+		rootCertExpireTime, err := nodeagentutil.ParseCertAndGetExpiryTimestamp(caCert)
+		if err != nil {
+			secretFetcherLog.Warnf("skip loading secret. Kubernetes secret %v contains a root "+
+				"certificate that fails to parse: %v", resourceName, err)
+			return nil, nil, isCAOnlySecret
+		}
 
-	t := time.Now()
-	newCert, newKey, newRoot, valid := extractCertAndKey(scrt)
-	if !valid {
-		secretFetcherLog.Warnf("Secret object: %v has empty field, skip adding secret", resourceName)
-		return
+		cans := &model.SecretItem{
+			ResourceName:                  resourceName,
+			CreatedTime:                   t,
+			Version:                       t.String(),
+			RootCertOwnedByCompoundSecret: false,
+			RootCert:                      caCert,
+			ExpireTime:                    rootCertExpireTime,
+		}
+
+		return nil, cans, isCAOnlySecret
```

Comment:
All three cases returns nil first? Can it be removed and just return cans, isCAOnlySecret?